library/Icingadb/Common/Macros.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+/* Icinga DB Web | (c) 2021 Icinga GmbH | GPLv2 */
+
+namespace Icinga\Module\Icingadb\Common;
+
+use Icinga\Application\Logger;
+use Icinga\Module\Icingadb\Model\Host;
+use Icinga\Module\Icingadb\Model\Service;
+
+trait Macros
+{
+    /**
+     * Known icinga macros
+     *
+     * @var array
+     */
+    protected $icingaMacros = [
+        'host.name'     => 'name',
+        'host.address'  => 'address',
+        'host.address6' => 'address6',
+        'service.name'  => 'name'
+    ];
+
+    /**
+     * Get the given string with macros being resolved
+     *
+     * @param string       $input  The string in which to look for macros
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function expandMacros($input, $object)
+    {
+        if (preg_match_all('@\$([^\$\s]+)\$@', $input, $matches)) {
+            foreach ($matches[1] as $key => $value) {
+                $newValue = $this->resolveMacro($value, $object);
+                if ($newValue !== $value) {
+                    $input = str_replace($matches[0][$key], $newValue, $input);
+                }
+            }
+        }
+
+        return $input;
+    }
+
+    /**
+     * Resolve a macro based on the given object
+     *
+     * @param string       $macro  The macro to resolve
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function resolveMacro($macro, $object)
+    {
+        $macroType = substr($macro, 0, strpos($macro, '.'));
+        if ($object instanceof Host) {
+            $objectType = 'host';
+        } else {
+            $objectType = 'service';
+        }
+
+        if (
+            isset($this->icingaMacros[$macro])
+            && isset($object->{$this->icingaMacros[$macro]})
+            && $macroType === $objectType
+        ) {
+            return $object->{$this->icingaMacros[$macro]};
+        }
+
+        try {
+            if (isset($this->icingaMacros[$macro]) && $objectType === 'service') {
+                // The macro does exist in the $this->icingaMacros though its type is different
+                $column = substr($macro, strpos($macro, '.') + 1);
```

Comment:
I guess `$this->icingaMacros[$macro]` should suffice here. This has two advantages:

* It's less work
* It's still correct even if the macro name is something entirely different than the expression